File: Array/Hard/cpp/3_Sum __Find_triplets_that_add_up_to_a_zero.cpp
//
// Created by santosh on 1/7/26.
//


#include<bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<vector<int>> findTriplets(vector<int>& nums)
    {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;

        for (int i = 0; i < nums.size(); i++)
        {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int left = i + 1;
            int right = nums.size() - 1;

            while (left < right)
            {
                int tripletsum = nums[i] + nums[left] + nums[right];
                if (tripletsum == 0)
                {
                    ans.push_back({nums[i], nums[left], nums[right]});
                    while (left < right && nums[left] == nums[left + 1]) left++;
                    while (left < right && nums[right] == nums[right - 1]) right--;
                    left++;
                    right--;
                }
                else if (tripletsum > 0)
                {
                    right--;
                }
                else
                {
                    left++;
                }
            }
        }

        return ans;
    }
};

int main()
{
    vector<int> nums = {-1, 0, 1, 2, -1, -4};
    Solution obj;
    const vector<vector<int>> ans = obj.findTriplets(nums);
    for (int i = 0; i < ans.size(); i++)
    {
        cout << "[";
        for (int j = 0; j < ans[i].size(); j++)
        {
            cout << ans[i][j] << " ";
        }
        cout << "]" << endl;
    }
}

File: Array/Hard/python /3_Sum_Find_triplets_that_add_up_to_a_zero.py
# 3Sum - Find triplets that add up to zero (LeetCode 15)
# Created by santosh

class Solution:
    def find_triplets(self, nums: list[int]) -> list[list[int]]:
        """
        Sort + two-pointer approach.
        Time: O(n^2), Space: O(1) (excluding output)
        """
        nums.sort()
        ans = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left = i + 1
            right = len(nums) - 1

            while left < right:
                triplet_sum = nums[i] + nums[left] + nums[right]

                if triplet_sum == 0:
                    ans.append([nums[i], nums[left], nums[right]])
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
                elif triplet_sum > 0:
                    right -= 1
                else:
                    left += 1

        return ans


if __name__ == "__main__":
    nums = [-1, 0, 1, 2, -1, -4]
    obj = Solution()
    ans = obj.find_triplets(nums)
    for triplet in ans:
        print(triplet)
